iree/compiler/Conversion/LinalgToSPIRV/ConcretizeTileAmongWorkgroupsPass.cpp

```diff
@@ -0,0 +1,540 @@
+class ConcretizeWorkgroupSizeOp final
+    : public OpRewritePattern<IREE::HAL::InterfaceWorkgroupSizeOp> {
+ public:
+  ConcretizeWorkgroupSizeOp(MLIRContext *context,
+                            SmallVector<int64_t, 4> workloadSize,
+                            SmallVector<int64_t, 4> tileSize,
+                            PatternBenefit benefit = 1)
+      : OpRewritePattern(context, benefit),
+        workloadSize(std::move(workloadSize)),
+        tileSize(std::move(tileSize)) {}
+
+  LogicalResult matchAndRewrite(IREE::HAL::InterfaceWorkgroupSizeOp op,
+                                PatternRewriter &rewriter) const override {
+    unsigned dimIndex = op.dimension().getZExtValue();
+
+    if (dimIndex < kWorkgroupDimCount) {
+      rewriter.replaceOpWithNewOp<ConstantOp>(
+          op, rewriter.getIndexAttr(tileSize[dimIndex]));
+      return success();
+    }
+
+    return failure();
+  }
+
+ private:
+  SmallVector<int64_t, 4> workloadSize;
+  SmallVector<int64_t, 4> tileSize;
+};
+
+/// Replaces hal.interface.workgroup.count op with the constant value chosen
+/// from tiling scheme.
+class ConcretizeWorkgroupCountOp final
+    : public OpRewritePattern<IREE::HAL::InterfaceWorkgroupCountOp> {
+ public:
+  ConcretizeWorkgroupCountOp(MLIRContext *context,
+                             SmallVector<int64_t, 4> workloadSize,
+                             SmallVector<int64_t, 4> tileSize,
+                             PatternBenefit benefit = 1)
+      : OpRewritePattern(context, benefit),
+        workloadSize(workloadSize),
+        tileSize(std::move(tileSize)) {}
+
+  LogicalResult matchAndRewrite(IREE::HAL::InterfaceWorkgroupCountOp op,
+                                PatternRewriter &rewriter) const override {
+    unsigned dimIndex = op.dimension().getZExtValue();
+
+    if (dimIndex >= kWorkgroupDimCount) return failure();
+
+    int64_t dimSize = workloadSize[dimIndex];
+    int64_t dimTile = tileSize[dimIndex];
+
+    if (dimSize == ShapedType::kDynamicSize || dimTile == 0) return failure();
+
+    int64_t count = ceilDiv(dimSize, dimTile);
+    rewriter.replaceOpWithNewOp<ConstantOp>(op, rewriter.getIndexAttr(count));
+
+    return success();
+  }
+
+ private:
+  SmallVector<int64_t, 4> workloadSize;
+  SmallVector<int64_t, 4> tileSize;
+};
+
+// Canonicalizes away a trip-one scf.for loop by inlining its body and removing
+// the loop.
+//
+// This pattern is needed because in Flow abstract tiling and distribution we
+// will create scf.for loops that distribute workload cyclically. After
+// concretizing hal.interface.workgroup.* ops, these scf.for loops still remain,
+// and they will be of the form:
+//
+//   %lb = mul %workgroup_id_{x|y|z}, %cst_tile_size_{x|y|z}
+//   scf.for %iv = %lb to %cst_wokload_size_{x|y|z}
+//                 step %cst_workload_size_{x|y|z} { ... }
+//
+// Such scf.for loops can be inlined if %lb is smaller than upper bound.
+class RemoveTripOneLoop final : public OpRewritePattern<scf::ForOp> {
```

Comment:
In general this pattern is unsafe. This is implicitly assuming that the target device can support the number of workgroups you need to make this happen. If that is not the case, this will result in incorrect execution. Ideally the number of workgroups is the "maximum number of workgroups needed". We should try to keep the option to use lesser (even maybe just 1 workgroup) and still be able to execute correct code. @ThomasRaoux had implemented this previously too and we had the same discussion at that time. Is this really needed or is it an optimizations. If the latter, we can leave the pattern here, add a flag (specific to this pass) that removes the one-trip loops and see if it has a performance impact. I am guessing it doesnt.

Reply:
Yes we need this. Without it, subviews used by conv inside the loop will have sizes computed from loop induction variables. Patterns for tiling and vectorizing conv won't kick in because of dynamic sizes in those subviews. So it's not simply an optimization. (Even if it's just an optimization it's worth to have it. Generally on the mobile world driver compilers aren't that reliable. Whatever we can do before/at SPIR-V level we should do to make sure we aren't bite by the underlying driver compiler not performing certain optimizations as we expect.)

Regarding unsafe b/c of the number of workgroups, I'm not sure I follow. What this pattern does is looking at the number of workgroups we will dispatch and check whether we surely have one trip. Given we have the loop upper bound equal to the loop step, as long as the loop lower bound is smaller than the loop upper bound, that's gonna be the case. That can be proved by checking the loop lower bound, which involves workgroup ID, against the workgroup count. The way we calculate the total number of workgroups is exactly the same as how we would do it on the host side. 

If the concern is that we might exceed the total number of workgroup allowed on the device, then I think the problem is not this pattern. It's actually how we perform the tiling. The tiling scheme and tile sizes determine the total number of workgroups we will dispatch, not this pattern. This pattern works inside a single tile (to be dispatched a single workgroup) after the tiling is done. It just simplifies the logic in one tile by canonicalizing trip-one loops away. Besides, the total number of workgroups is quite large in general. Even on Android, which is typically limiting, what we have there is a limit >= 65535 per dimension: [[0]](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount[0]&platform=android), [[1]](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount[1]&platform=android), [[2]](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount[2]&platform=android). While theoretically we might be exceeding it, in reality for the models we compile, it's hard to dispatch 65535^3 workgroups. ;-P

Reply:
This is rehashing a previous discussion again :D . I am not convinced that the loops need to be removed for tiling and vectorization to kick in. So if it is a correctness issue then we have a missing canonicalization.

To re-iterate from previous comment, the number of workgroups computed here is the "maximum number of workgroups" that would be needed, i.e. if you use more workgroups than needed, then those will not be used. The contract with the runtime which we would like to keep as long as it doesnt hurt performance is that the runtime can reduce the number of workgroups without having to recompile the executable. Removing the loops violates that contract. If you have to change the number of workgroups later then you need to recompile the executable. The loops here are just a few branches (that are convergent for the most part). So it should be cheap.

My understanding here is two things are being conflated. The removal of the loops makes some canonicalizations kick in that propogate the static shapes. Similar canonicalizations are needed to propogate the static shapes when the loop exists. Lets chat about this offline maybe.

Reply:
Discussed offline. For now will keep this as-is under a cl option so we can migrate all convolution vectorization patterns over to the linalg on tensors path. But created #5034 to track and investigate a more general way.